include/seqan3/range/container/small_string.hpp

```diff
@@ -0,0 +1,265 @@
+ * The stored string can be accessed as either a std::string or a c-style string through the respective member functions.
+ * All operations are constexpr except the conversion functions.
+ *
+ * Internally the string stores an array of size `N+1` containing the `0`-byte to allow conversion to
+ * c-style strings. This also means that when creating an instance of seqan3::small_string from a string literal
+ * containing a `0`-byte the size() of this small_string instance is one less than that of the literal.
+ *
+ * Example:
+ *
+ * \include test/snippet/range/container/small_string.cpp
+ *
+ */
+template <size_t capacity_>
+class small_string : public small_vector<char, capacity_ + 1>
+{
+    //!\brief The parent class.
+    using base_t = small_vector<char, capacity_ + 1>;
+
+public:
+    /*!\name Associated types
+     * \{
+     */
+    using typename base_t::value_type;
+    using typename base_t::reference;
+    using typename base_t::const_reference;
+    using typename base_t::iterator;
+    using typename base_t::const_iterator;
+    using typename base_t::difference_type;
+    using typename base_t::size_type;
+    //!\}
+
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+    //!\brief Inherit the parent class constructors.
+    using base_t::base_t;
+    //!\brief Inherit the parent class member function assign.
+    using base_t::assign;
+
+    //!\brief The default constructor initializes the underlying data member with '\0's.
+    constexpr small_string() noexcept : base_t{}
+    {
+        for (size_t i = 0; i < capacity_ + 1; ++i)
+            this->data_[i] = '\0';
+    }
+
+    /*!\brief Construction from literal.
+     * \param _lit The literal to construct the string for.
+     *
+     * ### Exceptions
+     *
+     * No-throw guarantee.
+     */
+    template <size_t N>
+    explicit constexpr small_string(char const (&_lit)[N]) noexcept : small_string{}
+    {
+        static_assert(N <= capacity_ + 1, "Length of string literal exceeds capacity of small_string.");
+        assign(_lit);
+    }
+
+    /*!\brief Construction from char.
+     * \param c The character to construct the small_string for.
+     *
+     * ### Exceptions
+     *
+     * No-throw guarantee.
+     */
+    explicit constexpr small_string(char const c) noexcept : small_string{}
+    {
+        assign(1, c);
+    }
+
+    /*!\brief Assign from literal.
+     * \param _lit The literal to assign the string from.
+     *
+     * ### Complexity
+     *
+     * Linear in the size of _lit.
+     *
+     * ### Exceptions
+     *
+     * No-throw guarantee.
+     */
+    template <size_t N>
+    constexpr small_string & operator=(char const (&_lit)[N]) noexcept
+    {
+        static_assert(N <= capacity_ + 1, "Length of string literal exceeds capacity of small_string.");
+        assign(_lit);
+        return *this;
+    }
+
+    /*!\brief Assign from literal.
+     * \param _lit The literal to assign the string from.
+     *
```

Comment:
Maybe add a sentence here that char arrays are expected to be null-terminated and that otherwise the last character is lost?